### PR TITLE
Update layout_for_admin component to take custom asset filenames 

### PR DESCRIPTION
**Pull request by davidgisbey**
## What

Allow custom JS and CSS filenames to be passed to the layout_for_admin component.

## Why

We're adding an admin interface to our application and want to use retain application.js & application.scss for the public interface.

This will allow us to pass admin.js and admin.scss in as args so we can retain only loading the required components for each so the page load time is fast.

## Visual Changes

N/A

## Testing

It was a bit awkward to test this as passing it a filename that doesn't exist causes exceptions. Using any_instance_of is a bit of a smell, but it's probs okay(ish) here to avoid additional complexity.

## Trello card

https://trello.com/c/r6uafy2D/1354-basic-admin-area

### CODE REVIEW

File: CHANGELOG.md
Comment (davidgisbey) [inline CHANGELOG.md:10]: I've added this to unreleased for now, but we might want to just bump the version to unblock us. Wdyt @kevindew?

Comment (andysellick) [reply CHANGELOG.md:10]: For preference we'd follow the normal release process: https://github.com/alphagov/govuk_publishing_components/blob/main/docs/publishing-to-rubygems.md

Comment (kevindew) [reply CHANGELOG.md:10]: Yup agree on normal release process, can be part of next batch release

File: spec/components/layout_for_admin_spec.rb
Comment (kevindew) [inline spec/components/layout_for_admin_spec.rb:21]: Do we need to do some mocking? i.e. does it error without it?

If we do need it, I'd suggest we use an allow here, as this isn't actually an assertion we're making about js_filename, if not I'd just delete it.

(Ideally it won't be too long until modernizr can be removed - I think it's only for IE < 11)

**Review by kevindew — COMMENTED** [review]
Great stuff David 👍

File: spec/components/layout_for_admin_spec.rb
Comment (davidgisbey) [reply spec/components/layout_for_admin_spec.rb:21]: Yes unfortunately it blows up without t

Comment (kevindew) [reply spec/components/layout_for_admin_spec.rb:21]: Sure, can we change it to allow to indicate it's not one of our expectations?

Comment (davidgisbey) [reply spec/components/layout_for_admin_spec.rb:21]: oh right sorry i get what you mean 👍 sorry i thought you meant both

**Review by kevindew — APPROVED** [review]
Great, all the changes look good to me. Only thing I remembered to suggest is that we should add some information in the commit message about how we ended up committing the sin of using `(allow|expect)_any_instance_of` and how we ended up stubbing ActiveView::Base - as there is quite opaque.

It'd be good to get a nod from @andysellick or one of the frontend devs on our team before merging though

**Comment by davidgisbey** [comment]
Yea great shout. Sorry i was using CI to run my tests as i was having issues locally and forgot to update the commit message. I'll sort that now 👍

File: CHANGELOG.md
Comment (andysellick) [inline CHANGELOG.md:12]: Nitpick: can this match the PR name please?

File: app/views/govuk_publishing_components/components/docs/layout_for_admin.yml
Comment (andysellick) [inline app/views/govuk_publishing_components/components/docs/layout_for_admin.yml:31]: These examples pass the default value, would it be clearer to show this functionality in action?

File: spec/components/layout_for_admin_spec.rb
Comment (andysellick) [inline spec/components/layout_for_admin_spec.rb:32]: Interested to know how these tests work - I was expecting an `assert` or similar, how do they work?

File: spec/components/layout_for_admin_spec.rb
Comment (andysellick) [inline spec/components/layout_for_admin_spec.rb:21]: Again just interested to know how this works - why does the JS test have `allow_any_instance_of` and `expect_any_instance_of`, but the CSS test only has `expect_any_instance_of`?

**Review by andysellick — COMMENTED** [review]
Change looks good, just a few comments/questions 👍

File: app/views/govuk_publishing_components/components/docs/layout_for_admin.yml
Comment (davidgisbey) [reply app/views/govuk_publishing_components/components/docs/layout_for_admin.yml:31]: The problem with that is that i'd have to add a blank `admin` stylesheet, because it will actually try to compile the file. So it if update it to use "admin" it'll look for "admin.js" and blow up if it can't find it.

I figured this was the lesser of 2 evils, but let me know if you'd prefer another approach 👍

Comment (andysellick) [reply app/views/govuk_publishing_components/components/docs/layout_for_admin.yml:31]: Oh I see that makes sense. Tricky! In that case maybe add a description to these examples to clarify? Something like:

```
description: An alternative JS filename can be used in place of the default `application.js` if required (note that this cannot easily be demonstrated here).
```

File: spec/components/layout_for_admin_spec.rb
Comment (davidgisbey) [reply spec/components/layout_for_admin_spec.rb:21]: Similar to above, the reason we're unable to make an assertion on the page is that it will blow up while trying to render the component if it's unable to find the file based on filename passed into the component. 

Due to that we have to stub out the method that handles the compilation of the assets. It feels like in this situation the thing we really care about is that the correct file is being included.

How allow and expect is that they stub out the method for the class passed in and return "" (by default).
So essentially what happens here is that the layout makes the call to compile `admin.js` and we intercept that call and return an empty string. Blocking it from actually trying to compile and blowing up.

The reason that we need `allow_any_instance_of` in the JS test is due to there being a second method call [here](https://github.com/alphagov/govuk_publishing_components/pull/3993/files#diff-1ebac9542b105f130b13ae04925c85d875d4dc231391387ecb4206b359fc0c07R19) 

So two calls are made:

```
 <%= javascript_include_tag "govuk_publishing_components/vendor/modernizr" %>
```
then
```
 <%= javascript_include_tag "application" %>
```
Without the allow this line

```
 expect_any_instance_of(ActionView::Base).to receive(:javascript_include_tag).with("admin").once
```

blows up because the `javascript_include_tag` method has been stubbed, but only once with expectation that the argument will be "admin".

The output of the test is something similar to

```
expected  javascript_include_tag with the argument of "admin", but it was actually "govuk_publishing_components/vendor/modernizr"
```

The `allow` allows the method to be called with any argument any amount of times which ensures we don't see the error above.  While the expect is checking that the call is made with "admin" once which is the desired out come

Hopefully that makes sense.

Maybe it's clearer if i update the allow to be more explicit and do 

```
allow_any_instance_of(ActionView::Base).to receive(:javascript_include_tag).with("govuk_publishing_components/vendor/modernizr")
 expect_any_instance_of(ActionView::Base).to receive(:javascript_include_tag).with("admin").once
```

So it's clear that the allow is due to the method being called twice, once with a file we're not worried about with our change to the implementation.

Comment (davidgisbey) [reply spec/components/layout_for_admin_spec.rb:21]: There are other approaches we could use like dynamically creating an empty `admin.js` file before action before running the test, testing what's in the file, then deleting it after running the test, but that comes with it's own setbacks. Still complex, far slower etc.

But happy to chat it through if you'd like!

I'll go back and update the commit message with some of the above.

File: spec/components/layout_for_admin_spec.rb
Comment (davidgisbey) [reply spec/components/layout_for_admin_spec.rb:32]: I've wrote a lengthy (sorry about that) comment [here](https://github.com/alphagov/govuk_publishing_components/pull/3993#discussion_r1579178525) which hopefully addresses this.

Let me know if any of it doesn't make sense

File: CHANGELOG.md
Comment (davidgisbey) [reply CHANGELOG.md:12]: Have updated the PR  since i think the changelog was a little more concise

File: app/views/govuk_publishing_components/components/docs/layout_for_admin.yml
Comment (davidgisbey) [reply app/views/govuk_publishing_components/components/docs/layout_for_admin.yml:31]: great shout. Thanks Andy will sort that now

**Comment by davidgisbey** [comment]
@andysellick i've made the requested changes. Hopefully the above explanation makes sense. Let me know if you not and/or you want any other changes 👍

File: spec/components/layout_for_admin_spec.rb
Comment (andysellick) [reply spec/components/layout_for_admin_spec.rb]: Really interesting stuff, thanks for taking the time to explain 👍

**Comment by davidgisbey** [comment]
Have tested it's working as expected on the component guide so i'll merge if there's no objections

**Pull request merged**
